File: client/app/src/main/java/com/packingandmoving/surveyagent/model/Survey.kt
package com.packingandmoving.surveyagent.model

import kotlinx.serialization.Serializable

/**
 * A paginated list envelope (Page_X_ in openapi.json). `data` of every list endpoint.
 */
@Serializable
data class Page<T>(
    val items: List<T>,
    val total: Int,
    val limit: Int,
    val offset: Int,
)

/** SurveyCreate (POST /surveys). */
@Serializable
data class SurveyCreate(
    val name: String,
    val originAddress: String,
    val destinationAddress: String,
    val preferredDatetime: String? = null,
)

/**
 * SurveyResponse (openapi.json). Decimal fields are serialized by the backend as
 * strings (Python Decimal) — kept as String? here to avoid float rounding; parse at
 * display time.
 */
@Serializable
data class Survey(
    val id: String,
    val name: String,
    val originAddress: String,
    val destinationAddress: String,
    val customerId: String,
    val surveyorId: String?,
    val preferredDatetime: String?,
    val status: SurveyStatus,
    val totalVolumeEstimate: String?,
    val totalValueEstimate: String?,
    val createdAt: String,
    val updatedAt: String,
)

/** SurveyStatusResponse (GET /surveys/{id}/status). Render buttons from availableActions. */
@Serializable
data class SurveyStatusInfo(
    val id: String,
    val status: SurveyStatus,
    val availableActions: List<String>,
    val processingStage: String? = null,
)

/** ReanalyzeRequest (POST /surveys/{id}/reanalyze). mode = "all" | "new_only". */
@Serializable
data class ReanalyzeRequest(val mode: String)

/** ReanalyzeResponse (202 from POST /surveys/{id}/reanalyze). */
@Serializable
data class ReanalyzeResult(
    val surveyId: String,
    val status: SurveyStatus,
    val jobIds: List<String> = emptyList(),
)

/** Optional reason body for POST /surveys/{id}/cancel. */
@Serializable
data class CancelRequest(val reason: String? = null)

/** Required reason body for POST /surveys/{id}/reject. */
@Serializable
data class RejectRequest(val reason: String)

File: server/app/schemas/survey.py
"""Survey request/response schemas."""

from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import ProcessingStage, ReanalysisMode, SurveyStatus


class SurveyCreate(BaseModel):
    """Customer-supplied survey request."""

    name: str = Field(min_length=1, max_length=255)
    origin_address: str = Field(min_length=1, max_length=1000)
    destination_address: str = Field(min_length=1, max_length=1000)
    preferred_datetime: datetime | None = None


class SurveyResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    origin_address: str
    destination_address: str
    customer_id: uuid.UUID
    surveyor_id: uuid.UUID | None
    preferred_datetime: datetime | None
    status: SurveyStatus
    total_volume_estimate: Decimal | None
    total_value_estimate: Decimal | None
    created_at: datetime
    updated_at: datetime


class SurveyStatusResponse(BaseModel):
    """Lightweight workflow view: current status + what the caller can do next."""

    id: uuid.UUID
    status: SurveyStatus
    available_actions: list[str]
    # Observed pipeline stage while ``status`` is PROCESSING; null otherwise.
    processing_stage: ProcessingStage | None = None


class ReanalyzeRequest(BaseModel):
    """Re-run AI analysis on a survey under review."""

    mode: ReanalysisMode = ReanalysisMode.ALL


class ReanalyzeResponse(BaseModel):
    survey_id: uuid.UUID
    status: SurveyStatus
    job_ids: list[uuid.UUID]


class RejectRequest(BaseModel):
    """Customer's revision request — a reason is required so the surveyor knows
    what to change."""

    reason: str = Field(min_length=1, max_length=2000)


class CancelRequest(BaseModel):
    reason: str | None = Field(default=None, max_length=2000)
